lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2077,12 +2077,62 @@ bool normalizeWeights(Module *M, ConvolutionNode &CV,
   return true;
 }
 
+/// Gets Constant or returns nullptr if input is not Constant.
+/// Skips QuantizeNode if present.
+static Constant *getConstant(const NodeValue &NV) {
+  Node *N = NV.getNode();
+  if (isa<QuantizeNode>(N)) {
+    N = N->getNthInput(QuantizeNode::InputIdx);
+  }
+  return dyn_cast<Constant>(N);
+}
+
 bool OptimizeBatchNorm::run(Function *F, const CompilationContext &cctx) {
   LOG_SCOPE(F->getLogContext(), getName());
   bool changed = false;
   auto &nodes = F->getNodes();
   auto *M = F->getParent();
 
+  // For each node:
+  for (auto &node : nodes) {
+    // Remove BN if mean,var,eps,scale,beta values make it redundant as per
+    // expression (X - mean) * (1.0 / sqrt(var + eps)) * scale + bias.
+    if (auto *BN = dyn_cast<BatchNormalizationNode>(&node)) {
+      auto *scaleC = getConstant(BN->getScale());
+      auto *biasC = getConstant(BN->getBias());
+      auto *meanC = getConstant(BN->getMean());
+      auto *varC = getConstant(BN->getVar());
+      if (!(scaleC && biasC && meanC && varC)) {
+        continue;
+      }
+      float scale, bias, mean, var;
+      if ((!isUniformConstant<float>(*scaleC, scale)) ||
+          (!isUniformConstant<float>(*biasC, bias)) ||
+          (!isUniformConstant<float>(*meanC, mean)) ||
+          (!isUniformConstant<float>(*varC, var))) {
+        continue;
+      }
+      float eps = BN->getEpsilon();
+
+      // Redundancy check based on strict equalities.
+      if (scale == 1.0f && bias == 0.0f && mean == 0.0f &&
+          (var + eps == 1.0f)) {
+        BN->getResult().replaceAllUsesOfWith(BN->getInput());
+        changed = true;
+        continue;
+      }
+
+      // Relaxed redundancy check based on reduced BN expression so that A
+      // is 1.0 and B is 0.0  in Y = A*X + B where,
+      // A = scale * (1.0 / (sqrt(var + eps))
+      // B = (bias - mean * (1.0 / sqrt(var + eps)) * scale)
+      if (bias == mean && (sqrt(var + eps) == scale)) {
+        BN->getResult().replaceAllUsesOfWith(BN->getInput());
+        changed = true;
+      }
+    }
+  } // For all nodes in the graph.
+
   // For each node:
   for (auto &node : nodes) {
```

Comment:
This loop is still going to process BNs that are dead but haven't been removed by DCE. Either we should merge the new optimization loop you added into this one, or else move this into a separate pass from the existing pass (would prefer not to call DCE in the middle of the pass explicitly).

Reply:
Good point! Agree, will merge the loops.